Add AffineLoopInterchange pass in the Affine Dialect.

HarshVardhanKumar opened the pull request:
Added AffineLoopInterchange pass alongwith certain modifications to supporting files. Also added test cases for AffineLoopInterchange pass.

HarshVardhanKumar committed:
Add AffineLoopInterchange pass

Adds a loop interchange pass to the Affine dialect.
The pass works only for loopnests having rectangular iteration space and no Affine-If statements.
In all other cases, it detects the absence of the above conditions and simply bails out leaving that particular loopnest unchanged.
The pass works for both - imperfectly nested loop nests as well as perfectly nested ones.

bondhugula reviewed (changes requested):
Could you create a single commit based on the upstream master? Avoid merge commits - always rebase.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,140 @@
+loopnest to perfectly nested loopnest. The process is described
+in next section. Once all the loop nests are converted to perfectly 
+nested loops, it starts dealing with each one of them in turn:
+*  First, it detects a loopnest and stores all the AffineForOps 
+involved in the loopnest in a llvm::SmallVector. Along the way, 
+it assigns loop IDs to each induction variable of the AffineForOp 
+it encounters. The loop IDs are assigned in such a way that the outermost 
+AffineForOp induction variable has an ID of 0. The loop-var-id are helpful 
+in calculating the access matrices of all the load/store operations at a later stage.
+
+* Next, we iterate over all possible permutations of the loopnest 
+to determine those permutations which do not violate any dependencies
```
Please linewrap everything to 80 cols.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,292 @@
+  return
+}
+
+
```
Drop extra new lines - just one is enough.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,1029 @@
+  DenseMap<Operation *, SmallVector<std::vector<int64_t>, 0>>
+      loopAccessMatrices;
+
+  // For a MemRef access function Ax+B, this represents B.
```
All of the member variable comments should be triple `///` doc comments.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,1029 @@
+  int forOpAPosition = 0;
+  llvm::SmallSet<unsigned,8> siblingsIndices;
+  int parentPosition = 0;
+  int index = 0;
```
unsigned.

HarshVardhanKumar replied:
Done. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,1029 @@
+  // visited. We need this order to compare affine.for ops within
+  // `parentForOp` with their copy in `copyParentForOp`.
+  // Comparing forOp.getOperation() does not work in that case.
+  int forOpAPosition = 0;
```
unsigned.

HarshVardhanKumar replied:
Done. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,1029 @@
+      // Note the position of `forOpA` in walk order.
+      forOpAPosition = index; 
+    }
+    for (unsigned i = 0; i<siblings.size(); i++) {
```
Please run clang-format.

HarshVardhanKumar replied:
Done, thanks.

bondhugula reviewed (changes requested):
Please clang-format your patch. 
```
git clang-format HEAD~
```

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+
+  /// For each memref access function Ax+B, this is
+  /// the collection of all A's indexed by their
+  /// respective affine.load/affine.store operations.
```
Please use the entire 80 col width. Reflow.

HarshVardhanKumar replied:
Done. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+  this->constVector.clear();
+
+  unsigned numOps = loadAndStoreOps.size();
+  for (unsigned i = 0; (i < numOps); ++i) {
```
Why is the guard in parentheses? Drop them.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+    uint64_t noDims = map.getNumDims();
+    uint64_t noSymbols = map.getNumSymbols();
+
+    // Number of rows in an accessMatrix = Number of dimensions
+    // in the memref object.
```
Use width.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+        auto constant = mapResult.cast<AffineConstantExpr>();
+        constVector[l] = (constant.getValue());
+      } else {
+
```
Drop space.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+      // leave the Row matrix to be a vector of zeroes.
+
+      if (mapResult.getKind() == AffineExprKind::Constant) {
+        auto constant = mapResult.cast<AffineConstantExpr>();
```
constant -> constExpr.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+          switch (expr.getKind()) {
+          case AffineExprKind::Add: {
+
+            // Does this row need a modification?
+            // Please note that the Row should not
+            // be modified in case of a constant affine expr.
```
Reflow.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+            // Does this row need a modification?
+            // Please note that the Row should not
+            // be modified in case of a constant affine expr.
+            bool modifyRow = true;
```
modifyRow or rowModified?

HarshVardhanKumar replied:
Whether this row will be modified or not (which initially is full of zeros) is a future decision based on the value of this variable. Using rowModified gives a sense of past as if the row has been modified. Hence I think modifyRow is more suitable?

HarshVardhanKumar replied:
Replaced the variable name 'modifyRow' with 'isConstExpr'. It is more suitable since at last all we are doing here is to test whether a sub-expr is constant or not. In case of a constant, we would not modify the row.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+/// Returns true if this entire column of `matrix` is zero.
+static bool checkColumnIsZero(ArrayRef<SmallVector<int64_t, 4>> matrix,
+                              unsigned column) {
+  for (auto &row : matrix) {
```
Avoid `auto` - spell out the type.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+    AffineMap map;
+    if (AffineLoadOp loadOp = dyn_cast<AffineLoadOp>(srcAccess.opInst))
+      map = loadOp.getAffineMap();
+    else if (AffineStoreOp storeOp = dyn_cast<AffineStoreOp>(srcAccess.opInst))
```
dyn_cast -> cast since this is expected to be a store now.

HarshVardhanKumar replied:
Here in this case, we don't know for sure whether srcAccess.opInst is really an AffneLoadOp or AffineStoreOp. If we were to use cast, we would first need to use isa<> and then use cast<>. Also, since the loadOp or storeOp are not reused any where else in the function definition, using both isa<> and cast<> does not seem fruitful. Instead dyn_cast<> serves the purpose well.
Please let me know your views on this.

However there is a later part of code that does gets benefitted by using cast<> instead of dyn_cast<>, and will be modified in next commit.
Thanks.

HarshVardhanKumar replied:
Fixed. Thanks!

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+    SmallVector<int64_t, 4> constVector(mapResults.size());
+
+    for (unsigned l = 0; l < mapResults.size(); l++) {
+
```
Drop blank line.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+      AffineExpr mapResult = mapResults[l];
+
+      // A Row of Matrix A.
+      SmallVector<int64_t, 4> Row;
```
Row -> row. Please use camelBack.

HarshVardhanKumar replied:
Removed this variable. Was unnecessary after certain modifications.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+      // Instead, push the value in constVector and
+      // leave the Row matrix to be a vector of zeroes.
+
+      if (mapResult.getKind() == AffineExprKind::Constant) {
```
Please don't use `getKind()` - instead use `isa<AffineConstantExpr>`.

HarshVardhanKumar replied:
Had tried earlier... using `isa<AffineConstantExpr>(mapResult)` where mapResult is `AffineExpr` throws error `classof is not a member of ‘mlir::AffineConstantExpr'`.

HarshVardhanKumar replied:
Will replace this and next statement with dyn_cast<> and ternary operator.

HarshVardhanKumar replied:
dyn_cast<> also fails with the same error. This is because dyn_cast<> has the statement ` isa<X>(Val) ? cast<X>(Val) : nullptr`. It is because of `isa<>` that it gives the same error.

HarshVardhanKumar replied:
used `mapResult.isa<AffineConstantExpr>()`. Worked! Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+          // expressions.
+          switch (expr.getKind()) {
+          case AffineExprKind::Add: {
+
```
Drop blank line.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+
+            // Parse LHS part of affine access expr.
+            AffineExprKind lhskind = lhs.getKind();
+            if (lhskind == AffineExprKind::DimId) {
+              auto dim = lhs.cast<AffineDimExpr>();
+              lhsPosition = indVarIdMap[operands[dim.getPosition()]];
+            } else if (lhskind == AffineExprKind::SymbolId) {
+              auto symbol = lhs.cast<AffineSymbolExpr>();
+              lhsPosition = noDims + symbol.getPosition();
+            } else if (lhskind == AffineExprKind::Constant) {
```
Don't compare with ==. Instead use isa<..> unless you are using a switch.

HarshVardhanKumar replied:
Replaced with switch.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+            AffineExprKind rhskind = rhs.getKind();
+            modifyRow = true;
+
+            if (rhskind == AffineExprKind::DimId) {
+              auto dim = rhs.cast<AffineDimExpr>();
```
isa + cast should be replaced with dyn_cast. Here and below.

HarshVardhanKumar replied:
Replaced with switch.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+
+        // Start parsing the mapResult.
+        mapResult.walk([&](AffineExpr expr) {
+          // Each `expr` can in turn be a combination of other affine
+          // expressions.
+          switch (expr.getKind()) {
+          case AffineExprKind::Add: {
```
You can't have such a long lambda! Outline to a separate function.

HarshVardhanKumar replied:
Split it into a helper function.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+        continue;
+      oneChild = false;
+      separateSiblingAffineForOps(forOperations[loopNest.first],
+                                  loopNest.second[loopNest.second.size() - 1],
```
loopNest.second.back()

HarshVardhanKumar replied:
Done this one and others.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+    }
+    if (isLoadStore) {
+      auto elementType = memref.getElementType();
+      unsigned sizeInBits = 0;
+      if (elementType.isIntOrFloat())
+        sizeInBits = elementType.getIntOrFloatBitWidth();
+  
+      // Store element size to be later used to check whether two
+      // accesses are within a cache-line/elementSize distance apart
+      // for a useful locality.
+      this->elementSize[op] = llvm::divideCeil(sizeInBits, 8);
```
There is already a function available to compute the size in bytes. See elsewhere.

HarshVardhanKumar replied:
Now element size can be deduced directly from memrefType object. No need to get elementType object. However did not find any function to compute size in bytes - in MemRefType class and mlir::Types.h
Will look at other files.

HarshVardhanKumar replied:
Done. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+uint64_t getParallelismCost(ArrayRef<unsigned> perm,
+                            SmallVector<bool, 4> &loopCarriedDV,
+                            SmallVector<unsigned, 4> &iterCounts) {
+  unsigned totalcost = 0;
+  unsigned thisLoopcost = 1;
```
Use uint64_t.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+/// Calculates a representative cost of a permutation regarding parallelism on
+/// multicores. A permutation having more free outer loops gets a smaller cost.
+uint64_t getParallelismCost(ArrayRef<unsigned> perm,
+                            SmallVector<bool, 4> &loopCarriedDV,
+                            SmallVector<unsigned, 4> &iterCounts) {
```
Pass as `ArrayRef` for inputs / read-only arguments.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+  groupId[dstOp] = groupId[srcOp];
+}
+
+/// Groups load/store ops in `loadAndStoreOps` list as reference groups to 
+/// calculate group-reuse. Two ops are in the same reference group if they 
+/// exibit group-temporal or group-spatial reuse with respect to an 
+/// affine.for loop at index `innermostForOp` in original loop nest.
+///
+/// Please refer Steve Carr et. al for a detailed description.
+/// https://dl.acm.org/doi/abs/10.1145/195470.195557
```
Clear explain which parameters are the input and output of the function.

in original -> in the original

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+      }
+    }
+  }
+  this->referenceGroups = referenceGroups;
```
You shouldn't really be storing state this way. Can you make the function take an output parameter?

HarshVardhanKumar replied:
Removed this one and similar others. No need to use output parameter, since this is a member function, it can access protected members directly.

HarshVardhanKumar replied:
Update: Removed use of all state variables.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+    float trip =
+        (forOp.getConstantUpperBound() - forOp.getConstantLowerBound()) / step +
+        1;
+    long double totalCacheLineCount = 0;
```
Why is a count a `double` entity insetad of uint64_t?

HarshVardhanKumar replied:
Here, we try to generalize for the cases where `step` may not be 1. In that case, `step` should be kept float to avoid integer division in calculation of `trip`.  We can keep `stride` as uint64_t, but calculation of count entity `cacheLinesForThisOperation` requires `(trip*stride)/8` - again a double/float quantity. 

`totalCacheLineCount` is sum of `cacheLinesForThisOperation` ( a double) - hence this count entity is kept `double.`

For cases where we consider only `step` = 1 always, no double quantity is needed.

HarshVardhanKumar replied:
Update: Replaced all double/float entities with appropriate unsigned values. For small values of `step`, integer division and double division should be close enough. Hence behaviour should remain the same.
Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+        1;
+    long double totalCacheLineCount = 0;
+
+    for (llvm::SmallSet<Operation *, 8> group : this->referenceGroups) {
```
Why `this->` while in the same class method?

HarshVardhanKumar replied:
Removed.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+  // be useful in calculating spatial locality cost for each permutation.
+  getCacheLineAccessCounts(loopNest, loadAndStoreOps);
+
+  // Get all permutations which do not violate any dependency constraints.
```
dependency -> dependence 

everywhere.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+                          validPerms);
+
+  // Return size 0 SmallVector to indicate that no interchange exists.
+  if (validPerms.size() < 2)
+    return bestPerm;
```
Use an output argument on the function instead of returning SmallVector's which leads to too many copies in general.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+/// Returns the optimal permutation considering both - the locality cost as well
+/// as the parallelism cost. The permutation which has the lowest overall cost is
+/// returned. Returns size-0 permutation in case no valid interchange exists.
+SmallVector<unsigned, 4> LoopInterchange::getBestPermutation(
```
Avoid returning SmallVector's - instead use an output argument appended to arguments and pass by reference.

HarshVardhanKumar replied:
Modified the function behaviour. Now no function returns SmallVector. Also made few changes within the function to avoid unnecessary copying : instead of saving bestPermutation, now it saves only the index of the best Permutation.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+  // bestPerm[i]. But here bestPerm[i] maps to loop
+  // at depth 'i'. Hence, we need to prepare a custom `returnPerm`
+  // according to requirements of `permuteLoops()`.
+  SmallVector<unsigned, 4> returnPerm;
+  returnPerm.resize(bestPerm.size());
```
Combine in the ctor.

HarshVardhanKumar replied:
Replaced with much simpler code.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+                            DenseMap<Value, unsigned> &indVarIdMap) {
+  loopNest.clear();
+  indVarIdMap.clear();
+  this->loopIterationCounts.clear();
+  this->loopCarriedDependenceVector.clear();
+  this->numberOfCacheLinesAccessed.clear();
+  this->referenceGroups.clear();
+  this->elementSize.clear();
+  this->loopAccessMatrices.clear();
+  this->constVector.clear();
```
`this->` is never used to access member variables in the LLVM style.

HarshVardhanKumar replied:
Removed

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,964 @@
+              op.getStep());
+        }
+
+        SmallVector<unsigned, 4> bestPermutation =
+            getBestPermutation(loopNest, indVarIdMap);
```
Pass an output argument by reference to avoid copies.

HarshVardhanKumar replied:
Done.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,142 @@
+## Affine Loop Interchange
+
+The file AffineLoopInterchange.cpp implements a loop interchange pass in affine 
```
in the Affine dialect ...

HarshVardhanKumar replied:
Done.

bondhugula reviewed (changes requested):
This is a good improvement but still several things to revise. I'm still not done with just the style changes.

bondhugula commented:
Please fix the PR title to a meaningful one.

bondhugula commented on mlir/include/mlir/Dialect/Affine/Passes.h:
```diff
@@ -73,7 +73,7 @@ std::unique_ptr<OperationPass<FuncOp>>
 /// Overload relying on pass options for initialization.
 std::unique_ptr<OperationPass<FuncOp>> createSuperVectorizePass();
-
+std::unique_ptr<OperationPass<FuncOp>> createAffineLoopInterchangePass();
```
Doc comment please. See above for reference.

HarshVardhanKumar replied:
Done. Thanks.

bondhugula commented on mlir/include/mlir/Dialect/Affine/Passes.td:
```diff
@@ -48,6 +48,12 @@ def AffineLoopInvariantCodeMotion
 
+def AffineLoopInterchange
+    : FunctionPass<"affine-loop-interchange">{
+  let summary = "Performs loop interchange of all affine loop nests";
```
Mention loop interchange for what? Locality, parallelism?

HarshVardhanKumar replied:
Done. Thanks.

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Updated comments.

HarshVardhanKumar committed:
Added summary for createAffineLoopInterchange() function.

HarshVardhanKumar committed:
Added detail for createAffineLoopInterchangePass()

HarshVardhanKumar committed:
Word wrapped to 80

HarshVardhanKumar committed:
Made modifications in grammar.

HarshVardhanKumar committed:
Update AffineLoopInterchange.md

HarshVardhanKumar commented:
Failed rebase attempt led to partial update to PR. Hence had to rollback to first version. This also removed the merge (which itself was a mistake). Made manual changes. The current version of files is the final version.

HarshVardhanKumar committed:
Changes with grammar of comments.

HarshVardhanKumar committed:
Unnecessary comments.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+};
+}
+
+/// Returns True if any affine.if op found in the loop nest rooted at `forOp`
```
True -> true

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+    case AffineExprKind::Constant: {
+      constantVectorValue += rhs.cast<AffineConstantExpr>().getValue();
+      isConstSubExpr = true;
+    }
```
Missing remaining cases - use `if` if it's just these two.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+      constantVectorValue += rhs.cast<AffineConstantExpr>().getValue();
+      isConstSubExpr = true;
+    }
+    }
```
Likewise.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+    }
+    // RHS in this case is always a constant or a symbol. For a constant, we
+    // don't need to modify the access matrix.
+  }
```
Add a default `unreachable` here. See other code for reference.

HarshVardhanKumar replied:
Fixed. Thanks. Actually, this part is indeed reachable. But we don't need to process those cases. Removed the comment. Adding llvm_unreachable() would throw errors.

bondhugula replied:
Add a comment on the cases you don't care about or convert this to if/else? Switch may be misleading.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+  }
+}
+
+/// For a memref access function AX+B, it calculates both A and B and stores
```
'X' is not mentioned about.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+    // consider a default value.
+    elementsSize[op] = memrefType.hasStaticShape()
+                           ? getMemRefSizeInBytes(memrefType).getValue() /
+                                 memrefType.getNumElements()
+                           : 8;
```
Don't hardcode this way. Define a static constant symbol in the pass and document it.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+
+  SmallVector<unsigned, 4> permutation;
+  for (unsigned i = 0; i < noLoopsInNest; i++)
+    permutation.push_back(i);
```
use std::iota instead.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+  while (std::next_permutation(permutation.begin(), permutation.end()))
+    if (isValidLoopInterchangePermutation(perfectLoopNest, permutation))
+      validPermutations.push_back(permutation);
+}
```
This will explode on large loop nests and is a design issue. You need to make a tradeoff here.

HarshVardhanKumar replied:
Limited to checking permutations of only four innermost loops in case of big loop nests.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+/// `rootForOp`. Result is stored in  `loopCarriedDependenceVector`.
+static void getLoopCarriedDependenceVector(
+    AffineForOp &rootForOp, ArrayRef<Operation *> loadAndStoreOps,
+    SmallVector<bool, 4> &loopCarriedDependenceVector, unsigned loopNestSize) {
```
Output arguments should be trailing ones.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+  unsigned numOps = loadAndStoreOps.size();
+  // Resize `loopCarriedDependenceVector` to fit entire loop nest.
+  loopCarriedDependenceVector.resize(loopNestSize);
+  for (unsigned i = 0; i < numOps; ++i) {
```
use `i = 0, e = ....size(); i < ...` form.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+}
+
+/// Calculates loop-carried-dependence vector for this loop nest rooted at
+/// `rootForOp`. Result is stored in  `loopCarriedDependenceVector`.
```
Instead of the obvious `Result is stored in ...` say what the result means: `... is set to true if ...`.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+command line flag -affine-loop-interchange.
+
+
+### Working:
```
`Working` isn't meaningful. `Design` / `Cost Model`?

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+nest can be parallelized if this permutation is followed.We select the permutation 
+with the minimum cost as the best permutation.
+
+* The cost model needs few inputs to evaluate cost of each permutation - a list  
```
evaluate cost -> evaluate the cost

Numerous errors like these.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+the one with minimum cost as the final interchange permutation. It then calls the 
+`permuteLoops` function for carrying out loop interchange.
+
+### Conversion from Imperfectly nested to Perfectly nested loopnests:
```
Avoid unnecessary capitalization. Imperfectly Nested -> imperfectly nested

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+The final cost of a permutation is defined as the sum over all the individual 
+loopcosts.This approach works because it favours the permutations which have 
+more number of free outer loops.
+
```
Drop extra blank line.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+
+#### Conclusion:
+
+Thus, the permutation to be selected as the best permutation is depends on all - 
```
Grammatically broken -> ".. is depends on all"

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+
+However for cases which do not conflict between spatial reuse and parallelism, 
+the approach will always produce the best permutation. For all other cases, the 
+best permutation selected will depend upond the amount of locality reuse and 
```
Spell fix. Please do a spell check.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+
+#### Spatial Locality Cost:
+
+The Spatial locality cost is based on size of the locality of each 
```
Spatial -> spatial

Similar errors at other places. Avoid unnecessary capitalization.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+### Cost Model:
+The final cost of a permutation consists of a combination of three separate costs.
+These are Parallelism cost, Spatial locality cost and Temporal reuse cost.
+
+FinalCost = 100* ParallelismCost + SpatialLocalityCost + TemporalReuseCost
+
```
How/why is the factor '100'  chosen? Document it.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+entire loop nest for a given permutation.
+
+* Using all this information, it gets a cost for each permutation and selects 
+the one with minimum cost as the final interchange permutation. It then calls the 
```
-> with the minimum cost ...

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+maximal temporal reuse corresponds to a minimum cost.
+
+This approach works since over all the permutation, the one which allows for 
+maximum temporal reuse will receive minimum cost - even if spatial costs are 
```
Please proofread thoroughly.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,804 @@
+    AffineExpr rhs = mulExpr.getRHS();
+    unsigned dimIdPos = 0;
+    if (lhs.isa<AffineDimExpr>()) {
+      auto dim = lhs.cast<AffineDimExpr>();
```
dimExpr

bondhugula commented on mlir/include/mlir/Dialect/Affine/Passes.td:
```diff
@@ -48,6 +48,13 @@ def AffineLoopInvariantCodeMotion
 
+def AffineLoopInterchange
+    : FunctionPass<"affine-loop-interchange">{
+  let summary = "Performs loop interchange of all the affine loopnests to 
```
of all the affine loopnests -> on all affine loop nests

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula reviewed (changes requested):
Btw, where are the test cases?

bondhugula reviewed (commented):
Can you time this pass on a 8-d loop nest?

bondhugula commented:
> Can you time this pass on a 8-d loop nest?

But first you'll need to add test cases. There can't be a revision without test cases.

HarshVardhanKumar committed:
Merge branch 'affineloopinterchange' of https://github.com/HarshVardhanKumar/llvm-project into affineloopinterchange

HarshVardhanKumar committed:
Added testcases for AffineLoopInterchange pass

HarshVardhanKumar committed:
Fixed grammatical error.

HarshVardhanKumar committed:
Limited checking permutations of only four innermost loops.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,264 @@
+}
+
+// Interchanged for spatial locality.
+// CHECK:       affine.load %arg0[%arg1, %arg2] : memref<2048x2048xf64>
+// CHECK-NEXT:  affine.store %{{.*}}, %arg0[%arg1, %arg2] : memref<2048x2048xf64>
```
Avoid numbered arguments. Instead capture IVs and use captured values. Here and everywhere below.

HarshVardhanKumar replied:
Fixed. Thanks.

HarshVardhanKumar replied:
Replaced with captured IV variables. Thanks.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,264 @@
+      // CHECK-NEXT:  affine.store %{{.*}}, %arg0[%arg3, %arg4] : memref<2048x2048xf64>
+    }
+  }
+
```
Missing test cases with dynamic loop bounds.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,264 @@
+  // CHECK-NEXT: addf
+  // CHECK-NEXT: affine.store %{{.*}}, %arg2[%arg3, %arg5] : memref<2048x2048xf64>
+  return
+}
```
Missing test case where interchange is invalid due to dependences.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+Such permutations are called `validPermutations`. If only one valid permutation 
+is found, we do not need to proceed any further.
+
+* In case more than one permutations are valid, we may need to consider each of 
```
permutation is valid

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md:
```diff
@@ -0,0 +1,136 @@
+variable has an ID of 0. These IDs are helpful in calculating the access matrices 
+for load/store ops.
+
+* Next, it iterates over all the possible permutations of a loopnest to determine 
```
loopnest -> loop nest -- here and everywhere else.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,800 @@
+}
+
+/// Returns true if this entire column of `matrix` is zero.
+static bool checkColumnIsZero(ArrayRef<SmallVector<int64_t, 4>> matrix,
+                              unsigned column) {
```
This naming and API isn't really meaningful. This shouldn't really be called a matrix in the first place. Second, you could have just passed matrix[column] to this method or better just checked inline whether the SmallVector is all zero.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,800 @@
+/// Fills `row` with the coefficients of loopIVs in `expr`. Any constant terms
+/// encountered in `expr` are added to `constantVectorValue`. Every value in
+/// `operands` should be a loopIV or a terminal symbol.
+static void prepareCoeffientRow(AffineExpr &expr, SmallVector<int64_t, 4> &row,
```
All input `AffineExpr` should be just passed by value. Output parameters should appear last in the list of args.

bondhugula replied:
Make similar changes everywhere else.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,264 @@
+
+// CHECK:      affine.load %arg0[%arg3, %arg4] : memref<2048x2048xf64>
+// CHECK-NEXT: affine.load %arg1[%arg4, %arg5] : memref<2048x2048xf64>
+// CHECK-NEXT: affine.load %arg2[%arg3, %arg5] : memref<2048x2048xf64>
+// CHECK-NEXT: mulf %0, %1 : f64
+// CHECK-NEXT: addf %2, %3 : f64
```
Do not use numbers in test cases - they can't be maintained easily. Either capture values and compare or use %{{.*}} if you don't want to check for them.

HarshVardhanKumar replied:
Fixed. Thanks.

HarshVardhanKumar committed:
Fixed grammatical mistakes.

HarshVardhanKumar committed:
Added testcases 

Added testcases for dynamic loop bounds and invalid interchange due to dependences

HarshVardhanKumar committed:
Spatial cost model and other changes.

HarshVardhanKumar committed:
Worked on words.

HarshVardhanKumar committed:
Fixed summary.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+
+/// Fills `row` with the coefficients of loopIVs in `expr`. Any constant terms
+/// encountered in `expr` are added to `constantVectorValue`. Every value in
+/// `operands` should be a loopIV or a terminal symbol.
```
Could you increase the size of `row` by one to store the tranlsation instead of having a separate `constantVectorValue`?

HarshVardhanKumar replied:
Fixed. Now, the last column of an access matrix consists of values of b-vector in Ax+b access function.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+                                DenseMap<Value, unsigned> &loopIndexMap,
+                                int64_t &constantVectorValue,
+                                SmallVector<int64_t, 4> &row) {
+  // TODO: Implement support for terminal symbols in `expr`.
```
Can you set row to be of the right size? Are you assuming the caller has already set it to the right size? That wouldn't be a good assumption. You should set it here instead of asserting.

HarshVardhanKumar replied:
It is resized before this function is called.

bondhugula replied:
Please don't keep the resize outside - it should be here.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+  case AffineExprKind::Add: {
+    // Is this sub-expr a constant? If yes, no need to modify `row`. Start by
+    // assuming the sub-expr is not a constant.
+    bool isConstSubExpr = false;
```
It's not clear here whaty sub expr here is. Expand comment or pick a better name.

HarshVardhanKumar replied:
Explained using an elaborate comment. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+    if (isa<AffineLoadOp>(srcOp))
+      map = cast<AffineLoadOp>(srcOp).getAffineMap();
+    else if (isa<AffineStoreOp>(srcOp))
+      map = cast<AffineStoreOp>(srcOp).getAffineMap();
```
Use `dyn_cast`. 
```
else if  (auto loadOp = dyn_cast<AffineLoadOp>(srcOp))
```

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+    canonicalizeMapAndOperands(&map, &operands);
+    ArrayRef<AffineExpr> mapResults = map.getResults();
+    // Number of rows in an accessMatrix = Number of dimensions in the memref
+    // object. Number of Columns = (noDimIDs + noSymbols).
```
no -> num.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+}
+
+/// Converts all imperfectly nested loop nests in `funcOp` to perfectly
+/// nested loop nests by loop splitting.
```
It's not clear what this means. Expand further. In what cases can you do that? Did you mean loop distribution? What about validity of this?

HarshVardhanKumar replied:
Added an elaborate comment. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+
+      forOperations[op.getOperation()] = op;
+    });
+    // Separate one of the sibling at a time.
```
siblings

HarshVardhanKumar replied:
Fixed here and at other places. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+/// accessed by the ops in `loadAndStoreOps`. These will be later used to
+/// check if two accesses are within a cache_line_size/element_size distance
+/// apart for a useful locality.
+static void getElementsSize(SmallVector<Operation *, 8> &loadAndStoreOps,
```
ElementsSize or ElementSizes?

HarshVardhanKumar replied:
Fixed here and everwhere else. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+/// accessed by the ops in `loadAndStoreOps`. These will be later used to
+/// check if two accesses are within a cache_line_size/element_size distance
+/// apart for a useful locality.
+static void getElementsSize(SmallVector<Operation *, 8> &loadAndStoreOps,
```
`loadAndStoreOps` is an input to function -> ArrayRef.

HarshVardhanKumar replied:
Fixed here and everywhere else. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+                            DenseMap<Operation *, unsigned> &elementsSize) {
+  // Assumption: In cases where element size is difficult to obtain, assume a
+  // default value of 8 bytes.
+  constexpr unsigned defaultEltSize = 8;
```
This can be a static field in the Pass class.

HarshVardhanKumar replied:
Done. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+      AffineStoreOp storeOp = cast<AffineStoreOp>(*op);
+      memrefType = storeOp.getMemRefType();
+    }
+    elementsSize[op] = memrefType.hasStaticShape()
```
This is really the memref size and not `elementsSize` - the latter can be misleading.

HarshVardhanKumar replied:
The memref size is divided by the number of elements in the memref to get the element size provided the memref has a static shape. Otherwise, we go with the default value.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+    AffineForOp &rootForOp, ArrayRef<Operation *> loadAndStoreOps,
+    unsigned loopNestSize, SmallVector<bool, 4> &loopCarriedDependenceVector) {
+
+  // Resize the `loopCarriedDependenceVector` to fit entire loop nest.
```
Rephrase this - you aren't resizing a single vector.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+/// not they exhibit group-temporal or group-spatial reuse with respect to an
+/// affine.for op present at depth `innermostIndex` in the original loop nest.
+///
+/// Please refer Steve Carr et. al for a detailed description.
```
refer to

HarshVardhanKumar replied:
Fixed. Here and everywhere else. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+/// affine.for op present at depth `innermostIndex` in the original loop nest.
+///
+/// Please refer Steve Carr et. al for a detailed description.
+/// https://dl.acm.org/doi/abs/10.1145/195470.195557
```
Include the paper title.

HarshVardhanKumar replied:
Fixed. Thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+///
+/// Please refer Steve Carr et. al for a detailed description.
+/// https://dl.acm.org/doi/abs/10.1145/195470.195557
+static void buildReferenceGroups(
+    SmallVector<Operation *, 8> &loadAndStoreOps,
+    DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+        &loopAccessMatrices,
+    DenseMap<Operation *, SmallVector<int64_t, 4>> &constVector,
+    DenseMap<Operation *, unsigned> &elementsSize, unsigned maxDepth,
+    unsigned innermostIndex,
+    SmallVector<llvm::SmallSet<Operation *, 8>, 8> &referenceGroups) {
```
Inputs should be ArrayRef, outputs should be SmallVector ref. Anywhere as well - I won't repeat it at other places. Please scan through and fix.

HarshVardhanKumar replied:
Replaced all occurances. `separateSiblingLoops` however, needs a small vector in order to identify locations of loops in original as well as cloned copy. Using ArrayRef in this case will not work since we loop1.getOperation() and loop2.getOperation() to compare two affine.for ops.

Using ArrayRef makes the affine.for ops const and this prohibits using the getOperation() function.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,761 @@
+  // We start with all accesses having their own group. Thus, if an access is
+  // not part of any group-reuse, it still has it's own group. Doing this takes
+  // care of self-spatial reuse.
+  constexpr unsigned cache_line_size = 64;
```
Use LLVM/MLIR style naming - camel back. Please see style guide on MLIR page. Also, this could be a static member in the pass itself so that it's easy to find and later set. `kCacheLineSize`.

HarshVardhanKumar replied:
Done. Thanks.

HarshVardhanKumar committed:
Added option to select cache line size

HarshVardhanKumar committed:
Update Passes.td

HarshVardhanKumar committed:
Added support for command line options for cache line size.

Removed the need for separate 'constVector'. The last column of access matrix is the constant b-vector of Ax+b. Replaced all occurances of SmallVector<....>& with ArrayRef<..>. Only the function `separateSiblingLoops` needs a small vector to identify the locations of each loop in the loop nest.

HarshVardhanKumar committed:
Used captured variables in CHECK

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Updated comments.

HarshVardhanKumar committed:
Update AffineLoopInterchange.md

HarshVardhanKumar committed:
Fixed resize of coefficient row.

HarshVardhanKumar committed:
Fixed some test cases.

HarshVardhanKumar committed:
modified `getAccessMatrices()`. Removed unnecessary arg.

HarshVardhanKumar committed:
Fixed for loop upper bounds.

HarshVardhanKumar committed:
Fixed some function declarations.

HarshVardhanKumar committed:
Fixed comments and modified `getElementSizes()`

HarshVardhanKumar committed:
Fixed variable names.

HarshVardhanKumar commented:
Test on 8-d loop nest: 
Time taken was 0.090387s. 
The loop nest used for test was:  
`func @big_loop(%A: memref<2048x2048x2048x2048xf64>, %B: memref<2048x2048x2048xf64>) {`  
`  affine.for %i = 0 to 2048 {  `
`    affine.for %j = 0 to 2048 {  `
`      affine.for %k = 0 to 2048 {  `
`        affine.for %l = 0 to 2048 {  `
`          affine.for %m = 0 to 2048 {  `
`            affine.for %n = 0 to 2048 {  `
`              affine.for %o = 0 to 2048 {  `
`                affine.for %p = 0 to 2048 {  `
`                  %v0 = affine.load %B[%m, %n, %p] : memref<2048x2048x2048xf64>  `
`                  %v1 = affine.load %B[%m, %n+1, %p] : memref<2048x2048x2048xf64>  `
`                  %v2 = mulf %v0, %v1 : f64  `
`                  affine.store %v2, %A[%m, %n, %o, %p] : memref<2048x2048x2048x2048xf64>  `
`                }  `
`              }  `
`            }  `
`          }  `
`        }  `
`      }  `
`    }  `
`  }  `
`  return  `
`}`

bondhugula commented:
> Test on 8-d loop nest:
> Time taken was 0.090387s.
> The loop nest used for test was:
> 

Was it on the release build? 90ms would be a lot for a single loop nest.

bondhugula commented:
> > Test on 8-d loop nest:
> > Time taken was 0.090387s.
> > The loop nest used for test was:
> 
> Was it on the release build? 90ms would be a lot for a single loop nest.

And how did you time it? You should use `-print-pass-timing`.

HarshVardhanKumar committed:
Changed the whole design of the code. Runtime improved.

Limited to only three innermost significant loops.

HarshVardhanKumar commented:
> > Test on 8-d loop nest:
> > Time taken was 0.090387s.
> > The loop nest used for test was:
> 
> Was it on the release build? 90ms would be a lot for a single loop nest.

It was measured on the current codebase.

It depends a lot on the function `mlir::isValidLoopInterchangePermutation`. It is called for every permutation. For big loop nests (depth>=8), this function is a **bottleneck**. Each call to this function takes approx 18-22ms for a loop nest of size 8. However for smaller loop nests, it creates no such issues.

HarshVardhanKumar commented:
Modified the previous implementation. Now, for loop nests of depth > 4, we consider only the innermost three loops for interchange. Rest of the outer loops are left as usual. The run time in this case for **8-d loop nests** is now reduced to **19.2ms**. 

Please note that this runtime is only for large loop nests. For small/medium cases, the implementation takes much less time. For example, the **entire test file** included (consisting of 13 testcases) was successfully processed in less than **35ms**.

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Fixed loop nest size.

HarshVardhanKumar committed:
Need to update it back. Prev. update was crashing.

HarshVardhanKumar committed:
Implemented state variables. 

No need to pass ArrayRef everytime. This took a lot of extra cycles. Use the state variables directly.

HarshVardhanKumar committed:
Removed support for cacheLineSize option. No need for it.

Used the default cacheLineSize everywhere.

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Removed support for cacheLineSize option.

The code now uses the default cache line size everywhere.

HarshVardhanKumar committed:
Updated comments. Fixed all grammatical errors.

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Fixed grammatical errors.

HarshVardhanKumar committed:
UPDATE: For loop depths more than four , consider only three.

For loops upto depth upto four, consider all the loops. This assymetry is due to the performance of the `mlir::isValidLoopInterchangePermutation` method used. For loop nests upto having depth more than four, it becomes a bottleneck.

HarshVardhanKumar committed:
Updated parallelism cost model.

HarshVardhanKumar committed:
Updated spatial reuse cost model.

HarshVardhanKumar committed:
Fixed comments.

HarshVardhanKumar committed:
Updated comments.

HarshVardhanKumar committed:
Updated conversion from imperfectly nested to perfectly nested for loops.

HarshVardhanKumar committed:
Update AffineLoopInterchange.md

HarshVardhanKumar committed:
Update AffineLoopInterchange.cpp

HarshVardhanKumar committed:
Update AffineLoopInterchange.md

HarshVardhanKumar commented:
Updated the doc file.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,344 @@
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 2048 {
+      affine.for %k = 0 to 2048 {
+        %a = affine.load %A[%i, %k] : memref<2048x2048xf64>
```
I see no negative test cases for the case on non rectangular loop bounds.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,344 @@
+      affine.for %k = 0 to 2048 {
+        %a = affine.load %A[%i, %k] : memref<2048x2048xf64>
+        %b = affine.load %B[%k, %j] : memref<2048x2048xf64>
+        %ci = affine.load %C[%i, %j] : memref<2048x2048xf64>
```
All test cases have a constant upper bound. Please change some to use symbolic upper bounds.

bondhugula commented on mlir/test/Dialect/Affine/affine-loop-interchange.mlir:
```diff
@@ -0,0 +1,344 @@
+
+// -----
+
+// Interchange is invalid due to dependences
```
Nit: terminate with period.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  void runOnAffineLoopNest();
+
+private:
+  /// Default cache line size(in bytes). Useful for getting a measure of the
```
size(in -> size (in

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  /// locality of each loop in a given loop nest.
+  constexpr static unsigned kCacheLineSize = 64;
+
+  /// Default element size to be used if a memref does not have a static shape.
+  constexpr static unsigned kDefaultEltSize = 8;
```
Dynamic shape doesn't mean the element size is unknown. The elt size is always known as long as it's an int/float type. So this comment looks incorrect.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  /// Default element size to be used if a memref does not have a static shape.
+  constexpr static unsigned kDefaultEltSize = 8;
+
+  bool isRectangularAffineForLoopNest();
+
+  void getLoopCarriedDependenceVector();
+
+  void getAllLoadStores();
```
Can drop spaces between these decl's.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+          &loopAccessMatrices,
+      DenseMap<Operation *, unsigned> &elementSizes);
+
+  uint64_t getNumCacheLinesSpatialReuse(ArrayRef<unsigned> perm);
+
+  uint64_t getNumSyncs(ArrayRef<unsigned> perm);
+
```
Likewise.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+
+  uint64_t getNumCacheLinesTemporalReuse(
+      ArrayRef<unsigned> permutation,
+      DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+          &loopAccessMatrices,
+      uint64_t maxTemporalReuse);
```
Inputs should be passed by `const` reference.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  bool getBestPermutation(DenseMap<Value, unsigned> &loopIndexMap,
+                          SmallVector<unsigned, 4> &bestPerm);
+
+  // Loop Carried Dependence vector. A 'true' at index 'i' means that the loop
```
Carried -> carried, ..

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  // Iteration count of each loop in the loop nest.
+  SmallVector<unsigned, 4> loopIterationCounts;
+
+  // The loop nest.
+  SmallVector<AffineForOp, 4> loopVector;
```
Unclear comment.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  // The loop nest.
+  SmallVector<AffineForOp, 4> loopVector;
+
+  /// Number of cache lines accessed by each loop in the loop nest.
+  DenseMap<const AffineForOp *, uint64_t> cacheLinesAccessCounts;
```
These would be wrong. The `AffineForOp` is just a value type wrapper around the actual op. The same `Operation` could thus potentially have different `AffineForOp *`. You need to hash on `Operation *`. This also means you aren't getting any caching here.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  /// Number of cache lines accessed by each loop in the loop nest.
+  DenseMap<const AffineForOp *, uint64_t> cacheLinesAccessCounts;
+
+  // List of all load/store ops in the loop nest body.
```
Triple comment please.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+    AffineExpr rhs = addExpr.getRHS();
+    unsigned lhsPosition = 0;
+    unsigned rhsPosition = 0;
+    if (lhs.isa<AffineDimExpr>()) {
+      auto dimExpr = lhs.cast<AffineDimExpr>();
```
Use `dyn_cast`. A `cast` followed by an `isa` almost always means a `dyn_cast` could have been used.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+    // The `rhs` may be a constant expr. In that case, no need to update the
+    // `row`.
+    bool isConstRhs = false;
+    if (rhs.isa<AffineDimExpr>()) {
+      auto dimExpr = rhs.cast<AffineDimExpr>();
```
dyn_cast

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+    if (rhs.isa<AffineDimExpr>()) {
+      auto dimExpr = rhs.cast<AffineDimExpr>();
+      rhsPosition = loopIndexMap[operands[dimExpr.getPosition()]];
+    } else if (rhs.isa<AffineConstantExpr>()) {
+      row.back() += rhs.cast<AffineConstantExpr>().getValue();
```
dyn_cast - otherwise, there would be an extra check.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+    // in these cases.
+    auto modExpr = expr.cast<AffineBinaryOpExpr>();
+    AffineExpr lhs = modExpr.getLHS();
+    if (lhs.isa<AffineDimExpr>()) {
+      auto dimExpr = lhs.cast<AffineDimExpr>();
+      row[loopIndexMap[operands[dimExpr.getPosition()]]] = 1;
```
dyn_cast.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+/// and store ops in the loop body. Please note that each affine access can be 
+/// represented as a linear system Ax+b (A is the affine access matrix, x is the 
+/// vector of loopIVs and b is the constant-term vector). `loopIndexMap` holds 
+/// depth locations of each loopIV in the original loop order.
```
This isn't true when you have floordiv/mod in accesses. What's the strategy in those cases. The test cases are also missing anything with a mod/div.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+      // to walk it. Instead, add the value to the constant b-vector element and
+      // leave the row unchanged. The last column of an access matrix stores the
+      // b-vector.
+      if (mapResult.isa<AffineConstantExpr>()) {
+        auto constExpr = mapResult.cast<AffineConstantExpr>();
```
dyn_cast.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+      op.getOperation()->erase();
+  });
+  // Erase the `lastSibling` from the the original copy.
+  lastSibling.getOperation()->erase();
```
You don't need `getOperation()`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+/// parent left in the new structure. Each sibling receives a separate copy
+/// of the common parent. This process is repeated until each parent has only 
+/// one child left.
+void LoopInterchange::handleImperfectlyNestedAffineLoops(Operation &funcOp) {
```
Pass `Operation` by pointer for consistency.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  // Store the arrangement of all the for-loops in the `funcOp` body in a tree
+  // structure. This makes storing the parent-child relationship an easy task. 
+  DenseMap<Operation *, SmallVector<AffineForOp, 4>> forTree;
+  // A helper map for the `forTree`. Since `AffineForOp` cannot act as a for
+  // a DenseMap, we've to use a map to convert to and from an affine.for to an
+  // Operation* and vice-versa.
+  DenseMap<Operation *, AffineForOp> forOperations;
```
This is unnecessary. You can `dyn_cast` the `Operation *` to its `AffineForOp`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+  for (Operation *op : loadAndStoreOps) {
+    if (isa<AffineLoadOp>(op)) {
+      memRefType = cast<AffineLoadOp>(*op).getMemRefType();
+    } else if (isa<AffineStoreOp>(op)) {
+      memRefType = cast<AffineStoreOp>(*op).getMemRefType();
```
dyn_cast.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp:
```diff
@@ -0,0 +1,789 @@
+}
+
+/// Calculates the number of synchronizations needed in this loop permutation.
+/// Those permutations having dependence satisfied on inner loops require
```
having -> that have
